src/core/namespace.c

```diff
@@ -1034,7 +1034,7 @@ static int apply_mount(
                         /* Hmm, either the source or the destination are missing. Let's see if we can create the destination, then try again */
 
                         if (stat(what, &st) < 0)
-                                log_debug_errno(errno, "Mount point source '%s' is not accessible: %m", what);
+                                log_error_errno(errno, "Mount point source '%s' is not accessible: %m", what);
```

Comment:
So we have this rule that functions that are "library-like" should not log about their errors beyond LOG_DEBUG, and instead propagate errors to their callers so that they can log about this. Functions that are "main-program-like" otoh are the ones that should log with LOG_WARNING or LOG_ERR. This way we can avoid duplicate logging or no logging at all, because it should be mostly clear whether a function logs itself or let's the caller log. (Also see https://systemd.io/CODING_STYLE/#error-handling)

The line between what is a "library-like" function and what is a "main-program-like" function is blurry, it's not always clear in which category some function belongs. But if you look at all functions from src/core/namespace.c you will see that in all cases we debug log here, we never log above LOG_DEBUG, hence so far it was decided that these functions are "library-like", i.e. helpers for code in execute.c (which usually logs about the errors).

Now, this patch of yours, first of all shouldn't use LOG_ERR, since we ignore the error here (by the earlier rule mentioned above), but secondly, it should stick to the rule that this is "library-like" code, and not log above LOG_DEBUG.

I am not entirely sure why you want to raise the log level here though. mount() returns ENOENT if source or dest are not found, and we propagate the error for that, hence it shouldn't be necessary to log about the error that we couldn#t create the mount points, since the ENOENT is clear enough that the mounting didn't work since some file hsa been referenced that doesn't exist.

Sure, the error strings could always be clearer in that case, but they aren't totally useless, and we do debug log, so if one needs more info turning on debug logging should clear everything up already?

Reply:
(Apologies for interjecting, Alan and myself are in the same team)

This feedback is coming from a few months of large scale trial runs - internal customers sometimes set a lot of BindPaths options in the units they create, and then sometimes all the required source folders are not created for $reasons.
Yes, it is a silly mistake, and totally the user's fault, but the only thing they get out is something like "Failure: could not create namespace", which is really not understandable and not actionable by them - so our team, as the platform owners, get high priority tickets raised, and then someone has to take the box offline to go and look at what happened because the extracted logs don't have enough, and it's all quite a big drain of time and resources, all to find out that a mkdir wasn't issued by the application responsible for deploying these.

These are servers in a datacentre with lots of constraints - there's no sysadmin that can just log in, flip the log to debug level and watch the journal for a bit, it's quite more complex than that. Self-inflicted wounds of course, but all with good motive. And we can't run all the time with log level set to debug for obvious reasons.

So it would save _really_ a lot of time and resources if there was a log message visible by default - and it doesn't have to be error level, it can be a warning or an info - that clearly spells out _why_ the namespace creation failed, so that internal customers can act on it by themselves. The telemetry systems can match on these and redirect to the right team automatically - so instead of a generic platform error, something that can point in the direction of the orchestrator with accuracy would help tremendously. What we get at the moment is not enough to do that, it's too generic.

The _how_ that happens is of course fine to change as required by the maintainers - a more complicated patch that lets the caller here know exactly that a mount failed because the source doesn't exist (and which calls out the source directory explicitly), rather than printing directly from "library-like" functions, would be perfectly fine as well, if it fits better with the coding style guidelines.

Thanks for the review!

Reply:
In general the no-logging-from-library-functions rule makes a lot of sense, but it can be problematic when the callee has a lot of internal work to do and can fail at many points. The caller then only gets a single error code and can't say much more than "something failed", which obviously is not very useful for users. Looking at the code, the failures for which errors are raised will indeed result in duplicate logging: the user will first get the concrete message like "Mount point source '%s' is not accessible" and then the general message "Failure: could not create namespace". I think this is OK, since the two messages provide different pieces of information for the user: the first is what exactly failed, and the second is why this matters. And it is worthwhile to log the precise reason for unit failure at non-debug level.